filetypes.py

```diff
@@ -62,85 +71,101 @@ class PreProcessor(Filetype):
             raise(PreProcessorException(Exception))
 
-        self.set_version(majorVal, minorVal, patchVal)
-
-    def overwrite_version(self):
-        file_contents = []
-        with open(self.target_file, "r") as input:
-            for line in input:
-                file_contents.append(line)
-        with open(self.target_file, "w", encoding='utf-8') as output:
-            for line in file_contents:
-                matchObj = re.search(self.r_var, line, re.X)
-                if matchObj != None: # preprocessor name ends in VERSION_MAJOR|VERSION_MINOR|VERSION_PATCH
-                    if 'MAJOR' in line:
-                        line = self.replace_part(self.version[0], line)
-                    elif 'MINOR' in line:
-                        line = self.replace_part(self.version[1], line)
-                    elif 'PATCH' in line:
-                        line = self.replace_part(self.version[2], line)
-                output.write(line)
-
-    def replace_part(self, part_num, line):
-        line = re.sub(r'(\d+)', part_num, line)  # change contents
-        return line
+        self.version_number = SemanticVersionNumber(int(major_val),
+                                                    int(minor_val),
+                                                    int(patch_val))
+
+    def update_version_in_file(self):
+        # Open file for reading
+        with open(self.target_file, 'r', errors='ignore', encoding='utf-8') as input:
+            content = input.read()
+
+        # Build replacement strings
+        major_repl = self.__build_replacement_string(self._match_obj_major,
+                                                     self.version_number.major)
+        minor_repl = self.__build_replacement_string(self._match_obj_minor,
+                                                     self.version_number.minor)
+        patch_repl = self.__build_replacement_string(self._match_obj_patch,
+                                                     self.version_number.patch)
+
+        # Substitute with new updated versions
+        content = self._match_obj_major.re.sub(major_repl, content)
+        content = self._match_obj_minor.re.sub(minor_repl, content)
+        content = self._match_obj_patch.re.sub(patch_repl, content)
+
+        # Write modified contents back to file
+        with open(self.target_file, 'w', errors='ignore', encoding='utf-8') as input:
+            input.write(content)
+
+    def __build_replacement_string(self, match_obj, new_version):
+        return ('#define ' +
+                match_obj.group('varName') +
+                match_obj.group('nSpaces') + '(' +
+                str(new_version) +
+                match_obj.group('unsigned') + ')')
 
 
 class Doxy(Filetype):
     #   Define regex patterns
-    r_pattern = r"PROJECT_NUMBER\s*=\s*(?P<major>\d+)\.(?P<minor>\d+)?\.(?P<patch>\*|\d+)"
+    r_pattern = r"PROJECT_NUMBER\s?=\s?(?P<major>\d+)\.(?P<minor>\d+)?\.(?P<patch>\*|\d+)"
```

Comment:
This does not work if Doxy uses tabs between tokens

Reply:
By tabs between tokens, do you mean `PROJECT_NUMBER\t=\tmajor.minor.patch`?

Reply:
Yes so,
PROJECT_NUMBER &emsp; = &emsp; 1.0.0     <- doesnt work
vs
PROJECT_NUMBER = 1.0.0 <- works
vs
PROJECT_NUMBER=1.0.0 <- works

Reply:
Okay, I figured it out. The first one you mentioned doesn't just have tabs in-between the equal sign, it has tabs surrounded by spaces:
`PROJECT_NUMBER \t = \t 1.0.0`

The question mark in `\s?` means 'zero or one' white-space.. but we have two in that particular case, so no match. So, I'm not sure if having tabs+spaces in-between the equal sign is something that we need to expect, but to make it work for all cases I'll change the `?` to a `*`.